### PR TITLE
regression test system bug fixes, eliminate MOM6 warnings (#2197), add xr_cnvcld flag to FV3 (#2185)

### LINKED ISSUE #2184: Address convection/cloud/radiation interaction at coarse resolution
## Description
GSL colleagues found a drift in SST in coupled forecasts done at C96 resolutions.

We identified this to be related to convection/cloud/radiation interaction - when suspended convective clouds are passed to the Xu-Randall cloud fraction and/or optical depth calculation we get this drift in SSTs in coupled simulations. The problem does not occur at C192, C384, C768 and C1152. 

## Solution

Add a namelist flag to control in which application this convection/cloud/radiation coupling should occur.

### LINKED ISSUE #2196: update MOM6 to eliminate compiling warning
Thanks for @marshallward, the updated code will eliminate compiling warning for MOM6 in UFS

### LINKED ISSUE #2199: Bug in atparse script
**What is wrong?**

It seems there's a mistake in the atparse script. The issue lies in how the while loop is structured within the atparse function. Specifically, the loop termination condition [[ "$__after" == "$__text" ]] may not always work as expected, leading to premature termination of the loop and missing the last line. So the last line in the file of diag_table_aod is always missing in the finale diag_table, which results in the missing of sea salt AOD in the diagnostic output.

**Steps to reproduce**

In the run directory, you can see that the last line of diag_table is always missing compared to the diag_table_template

**Additional information**

"gfs_phys", "DU_AOD_550", "du_aod550", "fv3_history2d", "all", .false., "none", 2
"gfs_phys", "SU_AOD_550", "su_aod550", "fv3_history2d", "all", .false., "none", 2
"gfs_phys", "BC_AOD_550", "bc_aod550", "fv3_history2d", "all", .false., "none", 2
"gfs_phys", "OC_AOD_550", "oc_aod550", "fv3_history2d", "all", .false., "none", 2
"gfs_phys", "SS_AOD_550", "ss_aod550", "fv3_history2d", "all", .false., "none", 2
The last line is always missing in the diag_table

**Do you have a proposed solution?**
In the atparse script, I've changed the while loop condition to while IFS= read -r __text || [ -n "$__text" ], which ensures that the loop continues until the input is fully consumed, even if the last line doesn't end with a newline character. It works well to solve the issue.

### LINKED ISSUE #2206: Empty partition tag causes some regression tests to fail on Hera Rocky using Rocoto
## Description
In Rocoto mode, rt.sh generates these lines on Hera in rocoto_workflow.xml:

```xml
<partition></partition>
```

An empty `<partition>` tag causes random failures on Hera Rocky.

### To Reproduce:
Run rt.sh in the head of develop on Hera Rocky in rocoto mode.

### LINKED ISSUE #2207: Inconsistent logging in rt.sh Rocoto mode
## Description
The destinations of stderr and stdout are inconsistent between rt.sh modes.

1. With ecFlow, stdout and stderr go to the compile*.log files, but with Rocoto only stdout goes there.
2. Rocoto sends stderr to the test job log files, but not stdout (that file is empty).
3. Rocoto combines the stdout and stderr streams of compiles into a .log file, but splits the test jobs into .err and (empty) .out files.

I suggest:

1. Rocoto matches the ecFlow behavior and sends stdout and stderr together to compile*.log files.
2. Rocoto sends both stderr and stdout to job log files.
3. Rocoto combines stderr and stdout into one stream for job log files. This way, we don't lose the context of what errors go with what output. The separate out & err streams are in out & err files in the job's work directory if someone needs separate files.

### To Reproduce:
Run rt.sh in Rocoto and ecFlow modes.

### PR DESCRIPTION
**Pull request by SamuelTrahanNOAA**
## Commit Queue Requirements:

- [x] Fill out all sections of this template.
- [x] All sub component pull requests have been reviewed by their code managers.
- [x] Run the full Intel+GNU RT suite (compared to current baselines) on either Hera/Derecho/Hercules
- [x] Commit 'test_changes.list' from previous step
---
#### Part 1: atparse.bash

The `tests/atparse.bash` ignored the last line of the input if it lacked an end-of-line character (see #2199).

Now it will process such a line, and insert an end-of-line at the end if one was missing. (Some regression tests assume the output ends with an end-of-line.)

Also, I commented the file and updated the test program.

#### Part 2: Regression test system Rocoto fixes

A stray `<partition></partition>` on Hera rocoto_workflow.xml caused sporadic job failures on the new Hera Rocky nodes (#2206)

Inconsistencies in Rocoto logging have been corrected (see #2207)

And there has been some general clean-up in the tests/ directory, such as only loading rocoto or ecflow modules when needed and replacing tabs with eight spaces.

#### Parts 3 & 4: Combined two other pull requests to this one.

Pull requests #2185 and #2197 have been merged into this branch. Neither should change results. Preliminary testing on Hera and Hercules has shown no changes to baselines from this combined PR.

### Commit Message:

```
* UFSWM - atparse.bash: correctly handle input that doesn't end with an end-of-line character. Fix some bugs in Rocoto support and clean up rt.sh.
  * FV3 - namelist flag xr_cnvcld to control if suspended grid-mean convective cloud condensate should be included in cloud fraction and optical depth calculation in radiation in the GFS suite
    * ccpp - physics-level changes to implement new namelist variable
  * MOM6 - update MOM6 code to eliminate all compiler warnings
```

### Priority:

* Normal

### UFSWM:

* Fixes #2199 
* Fixed MOM6 issue [128](https://github.com/NOAA-EMC/MOM6/issues/128)
* Fixes #2206
* Fixes #2207 
* Fixes #2196
* Includes PR #2197
* Fixes #2184
* Includes PR #2185

### Sub component Pull Requests:

UFS-WM PR 2185
FV3: sub-PR [799](https://github.com/NOAA-EMC/fv3atm/pull/799)
ccpp-physics: sub-PR [184](https://github.com/ufs-community/ccpp-physics/pull/184)

UFS-WM PR 2197
MOM6 sub-PR [130](https://github.com/NOAA-EMC/MOM6/pull/130)

### UFSWM Blocking Dependencies:

* None

---
### Regression Test Changes (Please commit test_changes.list):

* No Baseline Changes.

### Input data Changes:

* None.

### Library Changes/Upgrades:

* No Updates
  
---

## Testing Log:
- RDHPCS
  - [x] Hera
  - [x] Orion
  - [x] Hercules
  - [x] Jet
  - [ ] Gaea
  - [x] Derecho
- WCOSS2
  - [x] Dogwood/Cactus
  - [ ] Acorn
- [x] CI
- [ ] opnReqTest (complete task if unnecessary)

### CODE REVIEW

**Comment by SamuelTrahanNOAA** [comment]
I'm running regression tests on Hercules now. Will report back when I have something interesting to say about them.

**Comment by BrianCurtis-NOAA** [comment]
@SamuelTrahanNOAA If you get some time, could you cleanup that script with: https://www.shellcheck.net/ i have work moving towards a GH action to check bash scripts, and am working towards updating all scripts in UFSWM to comply.

I ask because I am not sure I 100% understand this file, which might come in handy to comply with shellcheck, and it seems you might.

If not, that's fine, it will be worked on either way.

**Comment by SamuelTrahanNOAA** [comment]
> I ask because I am not sure I 100% understand this file

Hopefully the comments I'm adding will help people understand it.

**Comment by SamuelTrahanNOAA** [comment]
I ran it through ShellCheck and found only one genuine error. ShellCheck's other complaints were limitations of its parser.

**Comment by BrianCurtis-NOAA** [comment]
> I ran it through ShellCheck and found only one genuine error. ShellCheck's other complaints were limitations of its parser.

Thanks!

**Comment by SamuelTrahanNOAA** [comment]
Some of the regression tests assume atparse always outputs a trailing end-of-line character. I'll update it to add the missing end-of-line.

**Comment by SamuelTrahanNOAA** [comment]
And now it works!

All tests passed on Hercules with no changes to baselines.

**Comment by jkbk2004** [comment]
@SamuelTrahanNOAA I always wonder why we use @[] for parsing parm namlist files. I mean why not ${}. Parsing with environment might be a handy approach to parse with simple bash script. Any idea?

**Comment by SamuelTrahanNOAA** [comment]
> I always wonder why we use @[] for parsing parm namlist files. I mean why not ${}. Parsing with environment might be a handy approach to parse with simple bash script. Any idea?

Using `${varname}` prevents you from parsing a shell script with atparse. The syntax must be easily identifiable from any other language it may parse. No languages that I know of use `@[varname]`, so that's what I went with.

**Comment by SamuelTrahanNOAA** [comment]
The full version of atparse is more capable. You can do things like `@[varname:-default value]`, there are if/else-if/else/end-if equivalents, and you can compare variables. The bash version only has `@[varname]`.

**Comment by BrianCurtis-NOAA** [comment]
Also curious, why are we using atparse? What was the need it met?

**Comment by zach1221** [comment]
Hi, @SamuelTrahanNOAA can your sync up your branch, please? We may end up combining this PR with another that we test today.

**Comment by SamuelTrahanNOAA** [comment]
> Also curious, why are we using atparse? What was the need it met?

We needed a macro expansion language that was available everywhere. At the time, there wasn't one we could use, so we repurposed this one from the old shell-based HWRF. It has the advantage of being pure bash, so no extra software is necessary.

If you can find a better macro expansion language, we could switch to it. A global regular expression replacement could change `@[varname]` to the new language's format easily.

EDIT: The full version of atparse is written in Python and is in the HAFS. They're looking for alternatives for their next workflow system.

**Comment by SamuelTrahanNOAA** [comment]
Branch is up to date.

It would be great if @DusanJovic-NOAA could review this. He has expressed reservations about the script's readability in the past. Perhaps it is improved enough so people can understand it now.

**Review by DusanJovic-NOAA — DISMISSED** [review]
Thank you for fixing the bug, and more importantly for adding extensive comments.

**Comment by SamuelTrahanNOAA** [comment]
@DusanJovic-NOAA - When I corrected a comment, it dismissed your approval. Someone will need to re-approve this for it to be merged.

**Review by DusanJovic-NOAA — DISMISSED** [review]

**Comment by jkbk2004** [comment]
@sam can you combine in @lisa-bengtsson #2185 and @jiandewang #2197? need to point to the FV3 and mom6 branches. I was running #2185 for pre-test just to confirm how rocky8 behaves. Some jobs remain but looks pretty good. Testing develop branch as well. Looks ok. On hera, ecflow/5.11.4 is available. We need to update ecflow version. @zach1221 @FernandoAndrade-NOAA can you pick up from here? @BrianCurtis-NOAA FYI

**Comment by BrianCurtis-NOAA** [comment]
> @sam can you combine in @lisa-bengtsson #2185 and @jiandewang #2197? need to point to the FV3 and mom6 branches. I was running #2185 for pre-test just to confirm how rocky8 behaves. Some jobs remain but looks pretty good. Testing develop branch as well. Looks ok. On hera, ecflow/5.11.4 is available. We need to update ecflow version. @zach1221 @FernandoAndrade-NOAA can you pick up from here? @BrianCurtis-NOAA FYI

ecflow 5.11.4 is only available on hfe10 . abort_dep_tasks.py does not work with that. I'm not sure what the SA's need to do to get the ecflow python stuff working.

**Comment by jkbk2004** [comment]
right, @zach1221 @FernandoAndrade-NOAA we may use rocoto. But with ecflow version, we can avoid module loading complaints.
```
Jong.Kim@hfe05:/scratch2/NCEPDEV/marine/Jong.Kim/UFS-RT/rt-2185/tests$ module load ecflow/5.11.4
ECF_HOST = hfe10, ECF_PORT = 21954
```

**Comment by SamuelTrahanNOAA** [comment]
I've merged #2185 and #2197 into this branch.

**Comment by SamuelTrahanNOAA** [comment]
It seems the last commit broke Rocoto support on Hera by making the ecflow module mandatory:

```
Lmod has detected the following error: The following module(s) are unknown: "ecflow/5.5.3"

Please check the spelling or version number. Also try "module spider ..."
It is also possible your cache file is out-of-date; it may help to try:
  $ module --ignore_cache load "ecflow/5.5.3"

Also make sure that all modulefiles written in TCL start with the string #%Module
```

**Comment by SamuelTrahanNOAA** [comment]
I'm testing a fix that only loads the rocoto or ecflow modules if the script is run in rocoto or ecflow mode, respectively.

**Comment by SamuelTrahanNOAA** [comment]
The fix to load rocoto/ecflow when needed appears to work, but we'll know for certain when this is tested on all platforms.

**Comment by DusanJovic-NOAA** [comment]
> > @sam can you combine in @lisa-bengtsson #2185 and @jiandewang #2197? need to point to the FV3 and mom6 branches. I was running #2185 for pre-test just to confirm how rocky8 behaves. Some jobs remain but looks pretty good. Testing develop branch as well. Looks ok. On hera, ecflow/5.11.4 is available. We need to update ecflow version. @zach1221 @FernandoAndrade-NOAA can you pick up from here? @BrianCurtis-NOAA FYI
> 
> ecflow 5.11.4 is only available on hfe10 . abort_dep_tasks.py does not work with that. I'm not sure what the SA's need to do to get the ecflow python stuff working.

ecflow python support was built with python 3.11

```
$ module show ecflow/5.11.4
-----------------------------------------------------------------------------------------
   /apps/modules/modulefiles/ecflow/5.11.4:
-----------------------------------------------------------------------------------------
prepend_path("PATH","/apps/ecflow/5.11.4/bin")
prepend_path("MANPATH","/apps/ecflow/5.11.4/man")
conflict("ecflow")
setenv("ECF_ROOT","/apps/ecflow/5.11.4")
prepend_path("PYTHONPATH","/apps/ecflow/5.11.4/lib/python3.11/site-packages/ecflow")
setenv("ECF_HOSTFILE","/apps/ecflow/5.11.4/share/ecflow/etc/hostfile")
setenv("ECF_HOST","hfe10")
setenv("ECF_PORT","13679")
LmodMsgRaw("ECF_HOST = hfe10, ECF_PORT = 13679
")
```

Default python on Hera is 3.6. Python script that imports ecflow, I think, must be executed with 3.11. ecflow is compiled with stack (run ldd on /apps/ecflow/5.11.4/bin/ecflow_client to confirm). Python in that spack installation is 3.11. This one:
```
/apps/spack/linux-rocky8-x86_64/gcc-13.2.0/python-3.11.6-6gktyo4cce4mwejpb42dkucgevsaq5ja/bin/python3.11
```
I do not see the module to load that python version. They need to, either:
1) provide the module to load python 3.11
2) add dependency in ecflow module to correct python module
3) prepend path to that python bin directory
4) rebuild ecflow using system python 3.6

Also, they need to fix hostfile and ECF_HOST environment variable on all nodes (except hfe10). I see now that they probably want us to run on hfe10 as a dedicated node. 

Basically, they need to properly install ecflow.

**Comment by SamuelTrahanNOAA** [comment]
The `regional_atmaq_debug_intel` hit its time limit on Hera Rocky. I'm resubmitting it.

**Comment by SamuelTrahanNOAA** [comment]
The `regional_atmaq_debug_intel` hit its wallclock limit a second time on Hera Rocky.
It did this at *timestep 6*, so the model probably hung. Something is wrong.

I will ping @lisa-bengtsson and @jiandewang whose changes are in this PR. Do one of you know why this happened?

- HERA: `/scratch2/BMC/wrfruc/Samuel.Trahan/gsl-pr/atparse-fix/failed-jobs/regional_atmaq_debug_intel`

The `cpld_control_pdlib_p8_gnu` and `cpld_control_nowave_noaero_p8_gnu` also failed. The error looks like a network problem. Will resubmit those.

**Comment by SamuelTrahanNOAA** [comment]
The `cpld_control_pdlib_p8_gnu` hung twice due to a network error.

```
206: --------------------------------------------------------------------------
206: WARNING: Open MPI failed to TCP connect to a peer MPI process.  This
206: should not happen.
206:
206: Your Open MPI job may now hang or fail.
206:
206:   Local host: h20c54
206:   PID:        4189350
206:   Message:    connect() to 10.184.20.53:1051 failed
206:   Error:      Resource temporarily unavailable (11)
206: --------------------------------------------------------------------------
189: [h20c53][[36928,0],189][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(248) failed: Connection reset by peer (104)
189: [h20c53][[36928,0],189][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(255) failed: Connection reset by peer (104)
189: [h20c53][[36928,0],189][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(253) failed: Connection reset by peer (104)
189: [h20c53][[36928,0],189][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(232) failed: Connection reset by peer (104)
189: [h20c53][[36928,0],189][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(243) failed: Connection reset by peer (104)
189: [h20c53][[36928,0],189][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(250) failed: Connection reset by peer (104)
182: [h20c53][[36928,0],182][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(243) failed: Connection reset by peer (104)
192: [h20c53][[36928,0],192][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(232) failed: Connection reset by peer (104)
199: [h20c53][[36928,0],199][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(242) failed: Connection reset by peer (104)
182: [h20c53][[36928,0],182][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(232) failed: Connection reset by peer (104)
199: [h20c53][[36928,0],199][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(229) failed: Connection reset by peer (104)
199: [h20c53][[36928,0],199][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(231) failed: Connection reset by peer (104)
199: [h20c53][[36928,0],199][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(232) failed: Connection reset by peer (104)
192: [h20c53][[36928,0],192][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(215) failed: Connection reset by peer (104)
199: [h20c53][[36928,0],199][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(226) failed: Connection reset by peer (104)
194: [h20c53][[36928,0],194][btl_tcp.c:559:mca_btl_tcp_recv_blocking] recv(244) failed: Connection reset by peer (104)
206: [h20c54:4189350] 2 more processes have sent help message help-mpi-btl-tcp.txt / client connect fail
206: [h20c54:4189350] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
```

**Comment by SamuelTrahanNOAA** [comment]
The good news is: all tests passed on Hercules!

**Comment by jiandewang** [comment]
> The `regional_atmaq_debug_intel` hit its wallclock limit a second time on Hera Rocky. It did this at _timestep 6_, so the model probably hung. Something is wrong.
> 
> I will ping @lisa-bengtsson and @jiandewang whose changes are in this PR. Do one of you know why this happened?
> 
> * HERA: `/scratch2/BMC/wrfruc/Samuel.Trahan/gsl-pr/atparse-fix/failed-jobs/regional_atmaq_debug_intel`
> 
> The `cpld_control_pdlib_p8_gnu` and `cpld_control_nowave_noaero_p8_gnu` also failed. The error looks like a network problem. Will resubmit those.

PR2197 only involves cpld* jobs so regional job failure shall has nothing to do with it

**Comment by lisa-bengtsson** [comment]
> > The `regional_atmaq_debug_intel` hit its wallclock limit a second time on Hera Rocky. It did this at _timestep 6_, so the model probably hung. Something is wrong.
> > I will ping @lisa-bengtsson and @jiandewang whose changes are in this PR. Do one of you know why this happened?
> > 
> > * HERA: `/scratch2/BMC/wrfruc/Samuel.Trahan/gsl-pr/atparse-fix/failed-jobs/regional_atmaq_debug_intel`
> > 
> > The `cpld_control_pdlib_p8_gnu` and `cpld_control_nowave_noaero_p8_gnu` also failed. The error looks like a network problem. Will resubmit those.
> 
> PR2197 only involves cpld* jobs so regional job failure shall has nothing to do with it

@jkbk2004 ran pre-testing of PR2185 on Hera and it ran OK. It also looks like the regional_atmaq_debug_intel uses the  FV3_GFS_v15p2 physics suite, which should not be impacted by my code changes.

**Comment by jkbk2004** [comment]
I have no issue with regional_atmaq_debug_intel/cpld_control_pdlib_p8_gnu. Let me submit full test on hera.
[RegressionTests_hera.log](https://github.com/ufs-community/ufs-weather-model/files/14732839/RegressionTests_hera.log)

**Comment by SamuelTrahanNOAA** [comment]
I found the problem. This line in the Rocoto XML was breaking things on Hera Rocky:

```xml
    <partition></partition>
```

**Comment by SamuelTrahanNOAA** [comment]
I've opened bug reports for the Rocoto problems I've found. They're fixed in this branch. I'm rerunning rt.sh now to check that they work as intended.

**Comment by SamuelTrahanNOAA** [comment]
I've opened bug reports for the Rocoto problems I've found. They're fixed in this branch. My latest test of Hera Rocky with Rocoto worked.

EDIT: The last commit message should have read "merge remote" for hash 894c5f2, but it's too late to fix that now.

File: tests/rt.sh [hunk 1]
```diff
@@ -787,16 +807,22 @@ elif [[ $MACHINE_ID = jet ]]; then
 
 elif [[ $MACHINE_ID = s4 ]]; then
 
-  module load rocoto/1.3.2
-  module load ecflow/5.6.0
+  if [[ "${ROCOTO:-false}" == true ]] ; then
+    module load rocoto/1.3.2
+  fi
+  if [[ "${ECFLOW:-false}" == true ]] ; then
```
Comment (BrianCurtis-NOAA) [inline tests/rt.sh:828]: This should be elif. -r an -e won't be used at the same time.

Comment (SamuelTrahanNOAA) [reply tests/rt.sh:828]: To make that consistent across all the machines, I'd have a lot of "ifs" and "module loads" to rearrange. It would change the order of module loads.

Comment (DusanJovic-NOAA) [reply tests/rt.sh:828]: If we are adding conditional loading of rocoto and ecflow modules then all rocoto and ecflow variables should be moved in the corresponding if blocks. For example, in this section:
```
  module use /ncrc/proj/epic/rocoto/modulefiles
  if [[ "${ROCOTO:-false}" == true ]] ; then
    module load rocoto
  fi
  ROCOTORUN=$(which rocotorun)
  ROCOTOSTAT=$(which rocotostat)
  ROCOTOCOMPLETE=$(which rocotocomplete)
```
ROCOTORUN, ROCOTOSTAT and ROCOTOCOMPLETE should not be defined outside the 'if rocoto' block. Also, 'module use ....' should be moved.

File: tests/atparse.bash [hunk 2]
```diff
@@ -1,42 +1,106 @@
...
+        # non-empty (! -n "$__text"). This ensures we read the final line
+        # even if it lacks an end-of-line character.
+        if [[ $? -ne 0 ]] ; then
+            if [[ -n "$__text" ]] ; then
+                # Text remained, but it had no end-of-line.
+		:
+            else
+                break
+            fi
+        fi
```
Comment (DusanJovic-NOAA) [inline tests/atparse.bash]: This section of code looks like this in my editor:

image

See how ':' is not properly indented. This is because it's indented with tabs. Your editor interprets tab as 8 spaces, mine as 4. Please replace tabs with spaces.

Comment (SamuelTrahanNOAA) [reply tests/atparse.bash]: It seems my emacs settings are not properly carrying over to Hera Rocky.

Comment (SamuelTrahanNOAA) [reply tests/atparse.bash]: And GNU's new Savannah server is blocked by the firewall, so I can't update /contrib/emacs with a version that works on Rocky.

Comment (jkbk2004) [reply tests/atparse.bash]: @SamuelTrahanNOAA I may clean up and push  on my side.

Comment (SamuelTrahanNOAA) [reply tests/atparse.bash]: I replaced tabs with spaces in rt.sh and atparse.bash.

Comment (jkbk2004) [reply tests/atparse.bash]: Cool!

File: tests/rt.sh (hunk 1 above)
Comment (SamuelTrahanNOAA) [reply tests/rt.sh:828]: You're right. All I did was a global editor macro execution for the "module load". I'll go back through and do what you suggest.

Comment (SamuelTrahanNOAA) [reply tests/rt.sh:828]: I did it. Please read through and tell me if it looks correct.
This is a pretty extensive change though. I'm retesting Rocoto on Hercules and Hera to make sure it works. Only a few regression tests; not the whole suite.

File: tests/rt.sh [hunk 3]
```diff
@@ -610,65 +610,80 @@ echo "Machine: " $MACHINE_ID "    Account: " $ACCNR
...
-  mkdir -p ${ECF_COMDIR}
+  if [[ "${ECFLOW:-false}" == true ]] ; then
+    ECFLOW_START=${ECF_ROOT}/scripts/server_check.sh
+    export ECF_OUTPUTDIR=${PATHRT}/ecf_outputdir
+    export ECF_COMDIR=${PATHRT}/ecf_comdir
+    rm -rf ${ECF_OUTPUTDIR} ${ECF_COMDIR}
+    mkdir -p ${ECF_OUTPUTDIR}
+    mkdir -p ${ECF_COMDIR}
+  fi
   export colonifnco=":output"  # hack
```
Comment (DusanJovic-NOAA) [inline tests/rt.sh:625]: This line should be inside 'ecflow' if block. But don't bother with it now if you already started testing. Can be done later.

Comment (SamuelTrahanNOAA) [reply tests/rt.sh:625]: I don't know the ecflow support code well enough to know what is specific to it.

Yes, I've started Rocoto tests on Hera and Hercules.

Comment (jkbk2004) [reply tests/rt.sh:625]: @zach1221 @FernandoAndrade-NOAA we will cross check on other machines.

File: tests/rt.sh [hunk 4]
```diff
@@ -707,16 +726,20 @@ elif [[ $MACHINE_ID = orion ]]; then
...
-  ROCOTO_SCHEDULER=slurm
+  if [[ "${ROCOTO:-false}" == true ]] ; then
+    module load contrib rocoto
+    ROCOTORUN=$(which rocotorun)
+    ROCOTOSTAT=$(which rocotostat)
+    ROCOTOCOMPLETE=$(which rocotocomplete)
+    ROCOTO_SCHEDULER=slurm
+  fi
 
   module use /work/noaa/epic/role-epic/spack-stack/orion/modulefiles
```
Comment (DusanJovic-NOAA) [inline tests/rt.sh:737]: I'm not sure, but I believe this 'module use' is only needed for loading ecflow module from this specific location. Needs to be tested and moved inside ecflow if block. Later.

Comment (SamuelTrahanNOAA) [reply tests/rt.sh:737]: I suspect several of the "module use" statements are specific to one workflow manager, but I didn't want to take the risk of removing one of them.

File: tests/rt.sh (hunk 1 above)
Comment (SamuelTrahanNOAA) [reply tests/rt.sh:828]: My mini-test of a few regression tests with Rocoto passed on Hercules and Hera.

Comment (SamuelTrahanNOAA) [reply tests/rt.sh:828]: Out of paranoia, I'm testing on Jet and Orion too.

File: tests/rt.sh (hunk 3 above)
Comment (zach1221) [reply tests/rt.sh:625]: Derecho fails with Rocoto, saying the job cannot be submitted with missing node size information. So, do we need to add derecho node size to rt_utils I assume? Ecflow still works.

Comment (SamuelTrahanNOAA) [reply tests/rt.sh:625]: Did it work in Rocoto mode on Derecho before?

File: tests/rt.sh (hunk 1 above)
Comment (SamuelTrahanNOAA) [reply tests/rt.sh:828]: Jet and Orion mini-tests passed now.

File: tests/rt.sh (hunk 3 above)
Comment (SamuelTrahanNOAA) [reply tests/rt.sh:625]: It seems the variable is ROCOTO_NODESIZE:

```
${ROCOTO_NODESIZE:+<nodesize>$ROCOTO_NODESIZE</nodesize>}
```

Comment (FernandoAndrade-NOAA) [reply tests/rt.sh:625]: > @zach1221 @FernandoAndrade-NOAA we will cross check on other machines.

Just adding verification from my side that there were no issues running a small set of tests with `-r` on Gaea, Jet, Hercules, and Orion.

**Comment by SamuelTrahanNOAA** [comment]
I've updated the commit message in the PR description. This commit message should describe all three pull requests. Could @lisa-bengtsson and @jiandewang please read the commit message and provide changes they want?

**Comment by jiandewang** [comment]
Look good, thanks

**Comment by SamuelTrahanNOAA** [comment]
It appears @lisa-bengtsson is on leave until Friday, so it is unlikely we'll hear back from her.
The commit message has a sentence straight from her PR description, so the content should be correct. I think we can use it even if we don't hear back.

**Comment by jkbk2004** [comment]
> It appears @lisa-bengtsson is on leave until Friday, so it is unlikely we'll hear back from her. The commit message has a sentence straight from her PR description, so the content should be correct. I think we can use it even if we don't hear back.

Yes, @lisa-bengtsson OK'd to merge once all tests run successfully. The commit messages look good.

**Comment by SamuelTrahanNOAA** [comment]
I see two problems left:

The FV3 PR has not been reviewed: https://github.com/NOAA-EMC/fv3atm/pull/799. If there are problems, someone will need to edit the branch to fix them.

Once the CCPP PR is merged (https://github.com/ufs-community/ccpp-physics/pull/184) someone will need to update `.gitmodules` in the branch for the FV3 PR (https://github.com/NOAA-EMC/fv3atm/pull/799) before the FV3 PR can be merged.

**Comment by FernandoAndrade-NOAA** [comment]
FYI Gaea is currently very unresponsive today to the point that it is taking a long time to clone and does so incorrectly. I've opened up a support ticket with them to see what's going on.

**Comment by jkbk2004** [comment]
No baseline change! we can skip Gaea due to the system issue. We can follow up with Gaea in the next PR if there is issue.

**Comment by SamuelTrahanNOAA** [comment]
We are changing the rt.sh slightly, and it is possible that change will break GAEA. Unlikely, but possible. Consider running only a few of the regression tests on GAEA when it comes back to see if the regression test system works.

**Comment by BrianCurtis-NOAA** [comment]
Acorn needed an adjustment to the changes where the ecflow if statements were, i'll push them after I run testing. Give me time to get those tests running.

**Comment by BrianCurtis-NOAA** [comment]
There's still something not quite right with Acorn using ecflow. The problem i'm having happens with both the current dev branch and this branch, so it's not a UFSWM issue, yet, at least. It's not worth holding this PR up. Please skip Acorn.

**Comment by zach1221** [comment]
> There's still something not quite right with Acorn using ecflow. The problem i'm having happens with both the current dev branch and this branch, so it's not a UFSWM issue, yet, at least. It's not worth holding this PR up. Please skip Acorn.

Ok, I think we're skipping Gaea as well, due to continued latency. So, we can begin the final review/merging process.

**Comment by zach1221** [comment]
@SamuelTrahanNOAA Fv3atm and MOM6 PRs are merged.
https://github.com/NOAA-EMC/fv3atm/commit/1ba84102cb814ec377efae92264fceb317c24398
https://github.com/NOAA-EMC/MOM6/commit/ab7bd14d209592d55490e75dbfaa61cb4a62df97

**Review by jkbk2004 — APPROVED** [review]

**Review by BrianCurtis-NOAA — APPROVED** [review]
